sympy/core/add.py

```diff
@@ -1005,19 +1005,15 @@ def _eval_as_leading_term(self, x, cdir=0):
         except TypeError:
             return expr
 
-        new_expr = new_expr.factor() # To factor the expanded multinomials
-        iszero = new_expr.is_zero
-        if iszero is None:
-            new_expr = new_expr.simplify()
-            iszero = new_expr.is_zero
-        if iszero is True:
+        new_expr = new_expr.trigsimp().factor()
```

Comment:
I would recommend `new_expr = factor_terms(new_expr.trigsimp().cancel())`

Reply:
An issue with this - 
```py
In [1]: a, b, c, x = symbols('a b c x', positive=True)

In [2]: limit((a + 1)*x - sqrt((a + 1)**2*x**2 + b*x + c), x, oo)
Out[2]: 
      ⎛       ______________    ⎞
      ⎜      ╱  2               ⎟
∞⋅sign⎝a - ╲╱  a  + 2⋅a + 1  + 1⎠
```

Reply:
This highlights the importance of getting the expression into form *before* it enters the `limit` code. If `cse` were used on the expression before beginning and the assumptions copied to the substituted variable, the correct result is obtained in master:
```python
>>> cse(eq)
([(x0, a + 1)], [x*x0 - sqrt(b*x + c + x**2*x0**2)])
>>> r, e = _
>>> x0=r[0][0]
>>> _x0 = Symbol('x0',**r[0][1]._assumptions)
>>> limit(e[0].subs(x0, _x0), x, oo).subs(_x0, r[0][1])
-b/(2*(a + 1))
```
The core workings should not have to deal with all the different ways to write expressions more than necessary. Even a wide-acting call like `simplify` should be used with great reservation (and preferably not at all).